### ISSUE TITLE
simple dynamical systems weirdness

### ISSUE THREAD

**Issue by epaxon** [brian-team/brian2#626]
Hi,

I was trying to make some examples of simple 2D linear dynamical systems using brian2. Found some strange behavior/bugs. Here is the code:

```python
from brian2 import *

# dynamic parameters
a = 0.0 / second
b = 1.0 / second
c = -0.5 / second
d = -0.1 / second

eqs_2D = Equations("""
    dx/dt = a * x + b * y : 1
    dy/dt = c * x + d * y : 1
""")

defaultclock.dt = 0.1 * ms
run_time = 40 * second

# This is just to run several simulation in parallel
n_runs = 10
system = NeuronGroup(n_runs, model=eqs_2D, threshold='x>100', reset='x=0; y=0', refractory=0*ms)

# set the initial conditions
system.x = np.random.randn(n_runs)
system.y = np.random.randn(n_runs)

system_monitor = StateMonitor(system, True, record=True)

simple2D = Network()

simple2D.add(system)
simple2D.add(system_monitor)
simple2D.store('init')
simple2D.restore('init')
simple2D.run(run_time)
```

So, this works fine so long as I set [a,b,c,d] such that the system does not have any spirals (stable or unstable). When I set [a,b,c,d] to have spirals (i.e. like it is set right now, which is just damped harmonic oscillator), then the simulation does not work -- every x and y value after the initial condition in the system_monitor is nan. Saddle points and stable stars run fine.

I've tried a ton of stuff to fix this (changing the dt, changing the parameters), but whenever I get a dynamical system that has a spiral it returns only nan values.

I finally did something completely trivial to the equations:

```python
eqs_2D = Equations("""
    dx/dt = a * x + b * y : 1
    dy/dt = c * x + d * y : 1
    dz/dt = 0 / ms : 1
""")
```

and then it works perfectly!

```python
figure('x and y vs t')
subplot(211)
plot(system_monitor.t / second, system_monitor.x.T)
subplot(212)
plot(system_monitor.t / second, system_monitor.y.T)

figure('x vs y')
plot(system_monitor.x.T, system_monitor.y.T)
```

![image](https://cloud.githubusercontent.com/assets/6740515/12283227/7b613256-b959-11e5-8ef0-a66b8efa02b7.png)
![image](https://cloud.githubusercontent.com/assets/6740515/12283231/813978c8-b959-11e5-9a42-50592b85ede5.png)

**Comment by mstimberg** [comment]
Hi, thanks for reporting this. I am not sure this strictly qualifies as a bug in our numerical integration, but we should certainly try to make situations like this clearer and less surprising (it is also somewhat related to #554). By default, if you don't specify any numerical integration method for your `NeuronGroup`, Brian tries to solve your equations exactly (over a single time step). This is what it did here, it was able to symbolically solve the equations with `sympy`. Without introducing new constants, the solution is quite complex, though: https://www.wolframalpha.com/input/?i={x%28t%29%27+%3D+a*x%28t%29+%2B+b*y%28t%29%2C+y%28t%29%27+%3D+c*x%28t%29+%2B+d*y%28t%29}

The problem with the solution is that depending on the values of the constants, the radicands can become negative -- we don't support complex arithmetics so this will result in not-a-number. Now, why does this change when you introduce `dz/dt = 0 / ms : 1`? In this case, we are no longer able to solve the linear system (which *is* a bug, I guess) so we will use a simple forward Euler integration. 

Long story short: for this kind of system I would not rely on Brian's ability to solve the system explicitly, instead use a numerical solver (specify `euler`, `rk2`, or `rk4` as the `method` argument of `NeuronGroup`). Since your system is linear, `exponential_euler` would also be a good choice but the solution contains divisions by `a` and `d`, so this won't work when these constants are zero. To avoid such problems, it can be a good idea to include the constants directly as values in the equations, instead of as symbols (for now the integration algorithm always leaves constants as symbols in the code, we might change this in the future though: #480). The easiest way to do this is to write your equations as `eqs_2D = Equations("""...""", a=a, b=b, c=c, d=d)`. Writing that, I just found out that this won't work with `exponential_euler` currently due to a misguided, too conservative, check whether the algorithm can integrate the equations...

**Comment by thesamovar** [comment]
Does Sympy have the ability to add assumptions (like a>0, b<0, etc.)? If so, maybe there is some way of letting users specify this to allow for exact solutions in cases like this?

I'm also wondering if we ought to prioritise adding the runtime error checking (even if just for nans) into the 2.0 milestone?

**Comment by mstimberg** [comment]
> Does Sympy have the ability to add assumptions (like a>0, b<0, etc.)? If so, maybe there is some way of letting users specify this to allow for exact solutions in cases like this?

Sympy does have this ability, but I am not sure this is enough here: first, the relevant information in the above example is not whether the constants itself are positive but whether a²  - 2 a d + 4 b c + d² is. Secondly, even replacing the constants by their values as I suggested above doesn't make the linear integration work, the solution still contains square roots of negative numbers (consistent with the solution Wolfram alpha shows if I use explicit values, note the imaginary number i: https://www.wolframalpha.com/input/?i=solve+{x%28t%29%27+%3D+0.0*x%28t%29+%2B+1.0*y%28t%29%2C+y%28t%29%27+%3D+-0.5*x%28t%29+%2B+-0.1*y%28t%29})

> I'm also wondering if we ought to prioritise adding the runtime error checking (even if just for nans) into the 2.0 milestone?

Maybe, situations like this are confusing quite a few people I guess...

**Comment by epaxon** [comment]
Ah, ok that makes a lot of sense. I figured it had something to do with the
complex numbers, but didn't realize it tried to solve the equations in
different ways under the hood.

Is there a reason it can't handle complex numbers? Is this a sympy issue?

Thanks for your help!
Paxon

**Comment by mstimberg** [comment]
> Is there a reason it can't handle complex numbers? Is this a sympy issue?

No, simpy would be fine with it. The reason is that under the hood we generate code (for numpy/Cython or C++) and there we assume real values. In principle it would be possible to make it work with complex numbers, but the complex-aware datatypes/functions are generally slower so we decided against it.

**Comment by thesamovar** [comment]
Just out of interest, I went back to this and reran it. For numpy, I get nans, but for weave and cython I just get that everything goes to 0 after the first step. So I'm pushing this up to high priority bug for 2.0.

**Comment by thesamovar** [comment]
Another strange thing that I don't understand. If I remove the ``c*x`` term it won't even let me use the linear integrator!

**Comment by mstimberg** [comment]
> For numpy, I get nans, but for weave and cython I just get that everything goes to 0 after the first step.

Huh, really? I'm getting NaN for all targets. I had a quick look and if I am not mistaken, then the C standard does not strictly require that `sqrt(negative_number)` returns `NaN`, but that seems to be what all libraries (including on Windows) do by default. The correct thing to check would be `errno`, which should be set to a domain error. But out of curiosity, what *does* the following print for you?
```Python
weave.inline(r'printf("%f\n", sqrt(-9.0));')
```

> Another strange thing that I don't understand. If I remove the c*x term it won't even let me use the linear integrator!

Yes, I don't quite understand that either... The problem occurs when sympy tries to calculate the matrix exponential of `Matrix([[d*dt, 0], [b*dt, a*dt]])`, it raises a `NotImplementedError`: 
```
Can't evaluate eigenvector for eigenvalue dt*(a + d)/2 - dt*Abs(a - d)/2
```

**Comment by thesamovar** [comment]
```
In [4]: weave.inline(r'printf("%f\n", sqrt(-9.0));',compiler='msvc')
-1.#IND00

In [5]: weave.inline(r'printf("%f\n", 1.0/sqrt(-9.0));',compiler='msvc')
-1.#IND00

In [6]: weave.inline(r'return_val = sqrt(-9.0);',compiler='msvc')
Out[6]: nan
```

So it is giving nan here, but not for the integrator. I'll look into this in more detail soon.

The matrix exponential thing is weird though.

**Comment by thesamovar** [comment]
OK so I looked into this and it appears that it will only give a problem if you include the ``--ffast-math`` or ``/fp:fast`` compiler options. Since these are off by default, I suggest that we add a note to the bit where we mention that you might want to switch these options on saying that it can cause this problem.

Are we sure that Brian will always give nan results and never incorrect results using the linear integrator? If so, I think we can close this issue if we check for nans. If not, and we can find an example, that's a dramatic problem that might mean we need to switch to euler as the default in all cases.

**Comment by mstimberg** [comment]
> OK so I looked into this and it appears that it will only give a problem if you include the `--ffast-math` or `/fp:fast` compiler options. Since these are off by default, I suggest that we add a note to the bit where we mention that you might want to switch these options on saying that it can cause this problem.

Um, actually we are using `--ffast-math` by default since almost a year now :-/ (see #585).

> Are we sure that Brian will always give nan results and never incorrect results using the linear integrator? If so, I think we can close this issue if we check for nans. If not, and we can find an example, that's a dramatic problem that might mean we need to switch to euler as the default in all cases.

I thought so, because stuff like `sqrt(-1)` gives NaN on Linux, but as you showed in your comment above, this does not seem to be that clear on Windows. 

How about adding a check for `if (errno != 0)` at the end of weave/Cython templates? It should not make much of a difference speed-wise (since it is outside of the loop) and it would also catch similar errors introduced by user code.

**Comment by thesamovar** [comment]
OK, I'll look into this.

**Comment by thesamovar** [comment]
Actually what should we do if ``errno!=0``? Pass a message back to Python somehow and log a warning? What about standalone? We could just log at the very end if ``errno`` was set?

**Comment by thesamovar** [comment]
OK, so I've made some progress on this issue. It turns out that it doesn't ever return 0 for ``sqrt(-1)``, it always returns some sort of nan. The problem was that the threshold condition ``x>100`` was leading to a spike every step, and that caused a reset of the variables to 0.

However, that's not all. Normally ``nan>100`` should return false. And it does. Unless you're running MSVC and both the ``/EHsc`` AND ``/fp:fast`` compiler flags are set.

Except it's more complicated than that. Because it still returns false, as you can see if you put a ``printf`` statement in the generated threshold code. However, as soon as you put that ``printf`` in the problem vanishes.

IT'S A GENUINE HEISENBUG! I'm so pleased. :)

So ``/EHsc`` is to do with exception handling, and ``/fp:fast`` does some funky things with floating point calculations which can raise exceptions I guess (like ``nan>100``). So I'm guessing that somehow if ``/EHsc`` is on and ``/fp:fast`` then somehow it can cause the exception to be mishandled in some incomprehensible way. I think this is probably a bug in the MS compiler, but I don't fully understand ``/EHsc`` so it may be my fault. In any case, I've changed the default options to not include ``/EHsc`` and I think this should eliminate this problem.

**Comment by thesamovar** [comment]
It's more complicated than I thought. Allowing ``/fp:fast`` and ``/Ox`` in standalone mode also has the same bug. So I'm going to switch off both ``/fp:fast`` and ``/EHsc`` in order to be safe. Unfortunately, this means the code will run a little slower than it might otherwise. I think we have to go for safety first though.

**Comment by mstimberg** [comment]
> Actually what should we do if errno!=0? Pass a message back to Python somehow and log a warning? What about standalone? We could just log at the very end if errno was set?

Why don't we raise an error directly? Is there any situation where you get an `errno` in a correctly running script?

**Comment by thesamovar** [comment]
No idea - worth testing.

**Comment by mstimberg** [comment]
I'd say we've addressed this with #750 (we now raise a warning in such a situation, hinting at a potential problem with the numerical integration algorithm). The situation is still far from ideal, we should make the linear integrator more safe (but this is already open as a separate issue: #679)
